### PR TITLE
Refactor nextflow code

### LINKED ISSUE #512: Check for unknown keys specified in `.run()`
For example, changing: 

https://github.com/viash-io/viash/blob/41eb44d63d7e8e4ec459cdf585e6f0915417960a/src/test/resources/testnextflowvdsl3/workflows/pipeline1/main.nf#L66-L69

to 

```groovy
 | step2.run( 
   // test map 
   map: { [ it[0], [ "input1" : it[1], "input2" : it[2] ] ] },
   brokkenpap: []
 ) 
```

should not work. Instead when I run `sbt "testOnly *Vdsl3ModuleTest"` the test succeed.

```
[info] welcome to sbt 1.6.1 (Ubuntu Java 11.0.20)
[info] loading settings for project viash-build-build-build from metals.sbt ...
[info] loading project definition from /home/di/code/viash/project/project/project
[info] loading settings for project viash-build-build from metals.sbt ...
[info] loading project definition from /home/di/code/viash/project/project
[success] Generated .bloop/viash-build-build.json
[success] Total time: 0 s, completed Aug 23, 2023, 10:06:24 AM
[info] loading settings for project viash-build from metals.sbt,plugins.sbt ...
[info] loading project definition from /home/di/code/viash/project
[success] Generated .bloop/viash-build.json
[success] Total time: 0 s, completed Aug 23, 2023, 10:06:24 AM
[info] loading settings for project viash from build.sbt ...
[info] set current project to viash (in build file:/home/di/code/viash/)
[info] Vdsl3ModuleTest:
[info] - Build pipeline components
[info] - Run pipeline
[info] - Test map/mapData/id arguments
[info] - Test fromState/toState arguments
[info] - Check whether --help is same as Viash's --help
[info] Run completed in 23 seconds, 90 milliseconds.
[info] Total number of tests run: 5
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 5, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[success] Total time: 24 s, completed Aug 23, 2023, 10:06:48 AM
```

### LINKED ISSUE #520: VDSL3: Add publish state functionality
If a workflow contains the following step, the events in the channel will be published at the end of the mycomponent.

```groovy
  | mycomponent.run(
    auto: [publish: "state"]
  )
```

if the publish dir is 'publishDir' and the channel contains `[myid, [output: "some.h5ad"]]`, the output directory is:

```
publishDir/
  myid.mycomponent.output.h5ad
  .myid.mycomponent.state.yaml
```

with `.myid.mycomponent.state.yaml`:

```yaml
id: myid
output: !!file myid.mycomponent.output.h5ad
```

## Id with slashes

unless 'id' contains a slash, then the dot is added after the last slash:

```
publishDir/
  foo/bar.mycomponent.output.h5ad
  foo/.bar.mycomponent.state.yaml
```

with `foo/.bar.mycomponent.state.yaml`:

```yaml
id: foo/bar
output: !!file bar.mycomponent.output.h5ad
```

### LINKED ISSUE #521: VDSL3: Set default directives tag to '$id'

### LINKED ISSUE #522: VDSL3: Set `simplifyOutput` to `false`
@tverbeiren Can we set this default to `false`? I don't like it when components behave differently depending on how many outputs the component has.

You can always set the default back to true with:

```
config_mods: |
  .platforms[.type == 'nextflow'].auto.simplifyOutput := true
```

### LINKED ISSUE #525: VDSL3: Detect when output IDs in a Nextflow workflow don't match up with the input IDs
When wrapping a Nextflow workflow, the output channel of the workflow is join with the input state in order to perform the `toState(id, output, state)` as well as find out about the desired filenames for files which are to be published.

However, this join will fail quietly if the user has performed an N-1 (`toList` or `toSortedList`), 1-N (`join`) operation:

    [f6/8d030d] process > wf:base:step1:processWf:step1_process (id)          [100%] 1 of 1 ✔
    [48/3b852d] process > wf:base:step2:processWf:step2_process (id)          [100%] 1 of 1 ✔
    [8e/4960ec] process > wf:base:step3:processWf:step3_process (newid)       [100%] 1 of 1 ✔
    [-        ] process > wf:publishStatesSimpleWf:publishStatesProc          -

When wrapping Nextflow workflows, the workflow should throw an error if the IDs of the output channel doesn't match the IDs of the input channel.

To resolve the issue, the user should store the original ID somewhere in the tuple.

Proposal: Store the original ID of the input tuple in the in the `_meta.join_id` field inside the state as follows:

  Example input event: `["id", [input: file(...)]]`,
  Example output event: `["newid", [output: file(...), _meta: [join_id: "id"]]]`

Example (N-1):

```groovy
    | toSortedList
    | map{ tups ->
      def new_id = "newid"
      def orig_id = tups[0][0]
      def new_state = [
        inputs: tups.collect{it[1].file},
        _meta: [join_id : orig_id]
      ]

      [ new_id, new_state ]
    }
```

### LINKED ISSUE #526: VDSL3: Automagically use `stageAs` to avoid input file name collision
Always use `stageAs` to avoid input file name collision in Nextflow processes:

```
  Process `wf:base:step3:processWf:step3_process` input file name collision -- There are multiple input files for each of the following file names: lines3.txt
```

### PR DESCRIPTION
**Pull request by rcannood**
### BREAKING CHANGES

* `NextflowPlatform`: Changed the default value of `auto.simplifyOutput` from `true` to `false` (#522, PR #518). With `simplifyOutput` set to `true`, the resulting Map could be simplified into a `File` or a `List[File]` depending on the number of outputs. To replicate the previous behaviour, add the following config mod to `_viash.yaml`:

  ```yaml
  config_mods: |
    .platforms[.type == 'nextflow'].auto.simplifyOutput := true
  ```

* `VDSL3Helper.nf`: Removed from the Viash jar file (PR #518). Its functions have been moved to `WorkflowHelper.nf`.

* `DataflowHelper.nf`: Added deprecation warning to functions from this file (PR #518).

* `preprocessInputs()` in `WorkflowHelper.nf`: Added deprecation warning to `preprocessInputs()` because this function causes a synchronisation event (PR #518).

### NEW FUNCTIONALITY

* `NextflowScript` & `NextflowPlatform`: Merged code for merging the `main.nf` files for VDSL3 components and wrapped Nextflow workflows (PR #518).
  By aligning the codebase for these two, wrapped Nextflow workflows are more similar to VDSL3 components. For example, you can override the behaviour of a
  wrapped Nextflow workflow using the `.run()` method. Status of a workflows `.run()` arguments:

  - Works as intended: `auto.simplifyInput`, `auto.simplifyOutput`, `fromState`, `toState`, `map`, `mapData`, `mapPassthrough`, `filter`, `auto.publish = "state"`
  - Does not work (yet): `auto.transcript`, `auto.publish = true`, `directives`, `debug`.

  In a next PR, each of the dependencies will have their values overridden by the arguments of the `.run`.

* `NextflowPlatform`: The data passed to the input of a component and produced as output by the component are now validated against the arguments defined in the Viash config (PR #518).

* `NextflowPlatform`: Use `stageAs` to allow duplicate filenames to be used automatigically (PR #518).

* `NextflowPlatform`: When wrapping Nextflow workflows, throw an error if the IDs of the output channel doesn't match the IDs of the input channel (#525, PR #518).
  If they don't, the workflow should store the original ID of the input tuple in the in the `_meta.join_id` field inside the state as follows:
  Example input event: `["id", [input: file(...)]]`,
  Example output event: `["newid", [output: file(...), _meta: [join_id: "id"]]]`

### MAJOR CHANGES

* `WorkflowHelper.nf`: The workflow helper was split into different helper files for each of the helper functions (PR #518).
  For now, these helper files are pasted together to recreate the `WorkflowHelper.nf`.

* `NextflowPlatform`: Set default tag to `"$id"` (#521, PR #518).

### MINOR CHANGES

* `NextflowPlatform`: Throw error when unexpected keys are passed to the `.run()` method (#512, PR #518).

## Issue ticket number and link

Closes #526
Closes #525
Closes #522
Closes #521
Closes #520
Closes #512

## Checklist before requesting a review
- [x] I have performed a self-review of my code

- Does this PR contain:
  - [x] Breaking changes
  - [x] New functionality
  - [x] Major changes
  - [x] Minor changes
  - [x] Bug fixes

- [x] Proposed changes are described in the CHANGELOG.md

- [x] Relevant unit tests have been added

### CODE REVIEW

File: src/main/resources/io/viash/platforms/nextflow/config/addGlobalParams.nf [hunk 1]
```diff
@@ -0,0 +1,54 @@
+// TODO: rename this to 'addGlobalArguments'
```
Comment (DriesSchaumont) [inline src/main/resources/io/viash/platforms/nextflow/config/addGlobalParams.nf:1]: Now may be a good time to do this TODO?

File: src/main/resources/io/viash/platforms/nextflow/workflowFactory/processDirectives.nf [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+  assert map instanceof Map : "Expected argument '$mapName' to be a Map. Found: class ${map.getClass()}"
+  map.forEach { key, val -> 
+    assert key in expectedKeys : "Unexpected key '$key' in ${mapName ? mapName + " " : ""}map"
+  }
+  requiredKeys.forEach { requiredKey -> 
+    assert map.containsKey(requiredKey) : "Missing required key '$key' in ${mapName ? mapName + " " : ""}map"
+  }
+}
+
+// TODO: unit test processDirectives
```
Comment (DriesSchaumont) [inline src/main/resources/io/viash/platforms/nextflow/workflowFactory/processDirectives.nf:11]: What about this TODO?

File: src/main/resources/io/viash/platforms/nextflow/workflowFactory/processFactory.nf [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+      // insert dummy into every output (see nextflow-io/nextflow#2678)
+      if (!par.multiple) {
+        ', path{[".exitcode", args.' + par.plainName + ']}'
+      } else {
+        ', path{[".exitcode"] + args.' + par.plainName + '}'
+      }
+    }
+    .join()
+
+  // TODO: move this functionality somewhere else?
```
Comment (DriesSchaumont) [inline src/main/resources/io/viash/platforms/nextflow/workflowFactory/processFactory.nf:78]: Since we are moving code around, what about this TODO?

**Review by DriesSchaumont — CHANGES_REQUESTED** [review]
Small questions/comments

File: src/main/resources/io/viash/platforms/nextflow/workflowFactory/processDirectives.nf (hunk 2 above)
Comment (rcannood) [reply src/main/resources/io/viash/platforms/nextflow/workflowFactory/processDirectives.nf:11]: This was already there. atm there is no infrastructure for unit testing individual groovy functions.

I can remove the comment, though.

File: src/main/resources/io/viash/platforms/nextflow/arguments/processInputsOutputs.nf [hunk 4]
```diff
@@ -0,0 +1,79 @@
+boolean typeCheck(String stage, Map par, Object x) {
```
[the remark this answers is not in the document]
Comment (rcannood) [reply src/main/resources/io/viash/platforms/nextflow/arguments/processInputsOutputs.nf:1]: k!

File: src/main/resources/io/viash/platforms/nextflow/workflowFactory/processFactory.nf (hunk 3 above)
Comment (rcannood) [reply src/main/resources/io/viash/platforms/nextflow/workflowFactory/processFactory.nf:78]: Hmmm... I'm going to pass on changing this code atm ^^

File: src/main/resources/io/viash/platforms/nextflow/config/addGlobalParams.nf (hunk 1 above)
Comment (rcannood) [reply src/main/resources/io/viash/platforms/nextflow/config/addGlobalParams.nf:1]: done!

**Review by Grifs — APPROVED** [review]
Scala code looks good to me

File: src/main/resources/io/viash/platforms/nextflow/workflowFactory/workflowFactory.nf [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+    // TODO: this join will fail if the keys changed during the innerWorkflowFactory
+    // join the output [id, output] with the previous state [id, state, ...]
+    out1_ = out0_.join(mid2_, failOnDuplicate: true)
+      // input tuple format: [id, output, prev_state, ...]
+      // output tuple format: [id, new_state, ...]
+      | map{
+        def new_state = processArgs.toState(it)
+        [it[0], new_state] + it.drop(3)
+      }
+      | _debug(processArgs, "output")
```
Comment (rcannood) [inline src/main/resources/io/viash/platforms/nextflow/workflowFactory/workflowFactory.nf:183]: I just realised this join will fail if the `innerWorkflowFactory` changes the ids (cc @tverbeiren: cfr that one discussion we had).

**Pull request merged**